driver/scsi_driver_extensions.c

```diff
@@ -277,6 +278,142 @@ WnbdHwResetBus(PVOID DeviceExtension,
     return TRUE;
 }
 
+UCHAR
+WnbdFirmwareInfo(PSCSI_REQUEST_BLOCK Srb)
+{
+    PSRB_IO_CONTROL SrbControl;
+    PFIRMWARE_REQUEST_BLOCK Request;
+    PSTORAGE_FIRMWARE_INFO_V2 Info;
+
+    SrbControl = (PSRB_IO_CONTROL)SrbGetDataBuffer(Srb);
+
+    Request = (PFIRMWARE_REQUEST_BLOCK)(SrbControl + 1);
+
+    if (sizeof(STORAGE_FIRMWARE_INFO_V2) > Request->DataBufferLength) {
+        SrbControl->ReturnCode = FIRMWARE_STATUS_INVALID_PARAMETER;
+        Srb->SrbStatus = SRB_STATUS_BAD_SRB_BLOCK_LENGTH;
+        goto exit;
+    }
+
+    Info = (PSTORAGE_FIRMWARE_INFO_V2)((PUCHAR)SrbControl + Request->DataBufferOffset);
+
+    if ((STORAGE_FIRMWARE_INFO_STRUCTURE_VERSION_V2 != Info->Version ) ||
+        (sizeof(STORAGE_FIRMWARE_INFO_V2) > Info->Size)) {
+        SrbControl->ReturnCode = FIRMWARE_STATUS_INVALID_PARAMETER;
+        Srb->SrbStatus = SRB_STATUS_BAD_SRB_BLOCK_LENGTH;
+        goto exit;
+    }
+
+    RtlZeroMemory((PCHAR)Info, Request->DataBufferLength);
+
+    /* Just fill data needed for WLK Firmware update test */
+    Info->Version = STORAGE_FIRMWARE_INFO_STRUCTURE_VERSION_V2;
+    Info->Size = sizeof(STORAGE_FIRMWARE_INFO_V2);
+    Info->UpgradeSupport = TRUE;
+    Info->SlotCount = 1;
```

Comment:
I was a bit confused about the slot field but then I found those docs: https://docs.microsoft.com/en-us/windows-hardware/drivers/storage/upgrading-firmware-for-an-nvme-device

So we're having a virtual device, not using any firmware. Can we skip tests that do not apply in our case instead of cluttering the code with unused stuff?

Reply:
I agree this is a bit mind numbing, but his applies just to one particular test on the 2019 playlist.

Unfortunately that test can't be skipped.

We could apply for an errata:
https://docs.microsoft.com/en-us/windows-hardware/test/hlk/user/windows-hardware-lab-kit-filters
but I have no idea about the resolution and how long that would take.

I added the implementation so we can have a clean run.

We can revert the commit later if we apply and get an errata for it.